pkg/controllers/subnetset/subnetport_handler.go

```diff
@@ -0,0 +1,91 @@
+package subnetset
+
+import (
+	"context"
+	"fmt"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+)
+
+// SubnetPortHandler supports lazy-creation of Subnet, the first Subnet won't
+// be created until there is a SubnetPort attached to it.
+// - SubnetPort creation: get available Subnet for the SubnetPort, create new
+//   Subnet if necessary.
+// - SubnetPort deletion: if recycling Subnet is required, delete Subnets without
+// SubnetPort attached to it.
+
+type SubnetPortHandler struct {
+	Reconciler *SubnetSetReconciler
+}
+
+// Create allocates Subnet for SubnetPort from SubnetSet.
+func (h *SubnetPortHandler) Create(e event.CreateEvent, _ workqueue.RateLimitingInterface) {
+	subnetPort := e.Object.(*v1alpha1.SubnetPort)
+	if subnetPort.Spec.Subnet != "" {
+		// Two possible scenarios:
+		// - 1. User uses `.Spec.Subnet` directly instead of `.Spec.SubnetSet`.
+		// - 2. Subnet has been allocated and `.Spec.Subnet` is rendered by SubnetPortHandler.
+		return
+	}
+	subnetSet := &v1alpha1.SubnetSet{}
+	key := types.NamespacedName{
+		Namespace: subnetPort.GetNamespace(),
+		Name:      subnetPort.Spec.SubnetSet,
+	}
+	if err := h.Reconciler.Client.Get(context.Background(), key, subnetSet); err != nil {
+		log.Error(err, "failed to get SubnetSet", "ns", key.Namespace, "name", key.Name)
+		return
+	}
+	log.Info("allocating Subnet for SubnetPort")
+	vpcList := &v1alpha1.VPCList{}
+	if err := h.Reconciler.Client.List(context.Background(), vpcList, client.InNamespace(subnetPort.GetNamespace())); err != nil {
+		log.Error(err, fmt.Sprintf("failed to get VPC under namespace: %s.\n", subnetPort.GetNamespace()))
+		return
+	}
+	vpcInfo, err := servicecommon.ParseVPCResourcePath(vpcList.Items[0].Status.NSXResourcePath)
+	if err != nil {
+		log.Error(err, "failed to resolve VPC info")
+		return
+	}
+	_, err = h.Reconciler.getAvailableSubnet(subnetSet, &vpcInfo)
```

Comment:
if new nsx subnet created for the first pod attached to the SubnetSet, we also need to update the SubnetSet status with the NSX subnet info.

Reply:
According to today's discussion, we may not need subnetport handler any more.
But you are right that we need to find a way to update subnetset status after a new subnet is created, will consider how to implement this, thanks.

Reply:
Done.